src/FFmpegWriter.cpp

```diff
@@ -511,17 +525,29 @@ void FFmpegWriter::WriteHeader() {
 	snprintf(oc->AV_FILENAME, sizeof(oc->AV_FILENAME), "%s", path.c_str());
 
 	// Write the stream header, if any
-	// TODO: add avoptions / parameters instead of NULL
 
 	// Add general metadata (if any)
 	for (std::map<string, string>::iterator iter = info.metadata.begin(); iter != info.metadata.end(); ++iter) {
 		av_dict_set(&oc->metadata, iter->first.c_str(), iter->second.c_str(), 0);
 	}
 
-	if (avformat_write_header(oc, NULL) != 0) {
+	// Set multiplexing parameters
+	AVDictionary *dict = NULL;
+
+	bool is_mp4 = strcmp(oc->oformat->name, "mp4");
+	bool is_mov = strcmp(oc->oformat->name, "mov");
+	// Set dictionary preset only for MP4 and MOV files
+	if (is_mp4 || is_mov)
+		av_dict_copy(&dict, mux_dict, 0);
+
+	if (avformat_write_header(oc, &dict) != 0) {
 		throw InvalidFile("Could not write header to file.", path);
 	};
 
+	// Free multiplexing dictionaries sets
+	if (dict) av_dict_free(&dict);
+	if (mux_dict) av_dict_free(&mux_dict);
```

Comment:
I'm also not sure about freeing `mux_dict` here. Should that be moved to `FFmpegWriter::Close`, maybe? If you're just going to free `mux_dict` right away here, then it really _is_ temporary storage and there's no reason not to pass it in to `avformat_write_header()` and let it get overwritten.

I get where the semantics get weird, because you want to pass in an `AVDictionary **` to `avformat_write_header()`, but you only want it to contain `mux_dict` IFF the export format is the right type. So, how about doing this?
```cpp
AVDictionary **dictp = NULL;
if (mux_dict && (is_mp4 || is_mov))
  dictp = &mux_dict;
if (avformat_write_header(oc, dictp) != 0) { ... }
if (mux_dict) av_dict_free(&mux_dict);
```
That way, you'll still pass a `NULL` into `avformat_write_header()` if there's no dict or it's the wrong format, same as before. If everything _is_ right to set the muxing options, then you'll pass in a pointer to the `mux_dict`, but without having to needlessly copy it first.

Reply:
(And in that case, you can disregard my previous comments about not labeling it temporary storage or adding a comment to explain the copying.)

Reply:
The dictionary only needed to pass the options into the function, nothing else. If there was other way to say to FFmpeg that I need faststart or fragmented writing with my MP4 export then code would be completely different. I don't know about other solutions. Maybe someone with better knowledge of FFmpeg and OpenShot can do it better.